### PR TITLE
Generate layer attributes on a worker (PR 1/3)

### PR DESCRIPTION
**Pull request by Pessimistress**
For #2489 

This PR contains necessary changes to the core classes.

### Change List
- Use the latest mjolnir.js (can be safely imported in a worker)
- Use the latest luma.gl (`Attribute` ignores missing WebGLContext, and properly infers index type)
- Make the following constructors ignore missing WebGLContext: `AttributeManager`, `AttributeTransitionManager`, `LayerManager`.
- `numInstances` and `vertexCount` defined in layer `props` override those in `state`.
- When using external buffers for attribute values, relax type check from throwing an error to a warning.

### CODE REVIEW

File: modules/core/src/lib/deck.js [hunk 1]
```diff
@@ -28,13 +28,20 @@ import log from '../utils/log';
...
+// import {EventManager} from 'mjolnir.js';
 
 import assert from '../utils/assert';
 import VENDOR_PREFIX from '../utils/css-vendor-prefix';
 import {EVENTS} from './constants';
 /* global document */
 
+let EventManager = null;
+try {
+  EventManager = require('mjolnir.js').EventManager;
```
Comment (Pessimistress) [inline modules/core/src/lib/deck.js]: This is rather unfortunate - when using the `browser` field in package.json to support node, it does not cover the worker case.

File: test/apps/attribute-worker/worker.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    if (ArrayBuffer.isView(attribute.value)) {
+      slimAttribute = attribute.value;
+      transferList.push(attribute.value.buffer);
+    } else {
+      slimAttribute = {
+        constant: attribute.constant,
+        value: attribute.value
+      };
+    }
+    props[attributeName] = slimAttribute;
```
Comment (Pessimistress) [inline test/apps/attribute-worker/worker.js]: @ib I wonder if there's a cheap way to pack this object into a binary? Imagine snapshooting a composite layer, generating the transfer list would be cumbersome.

File: modules/core/src/lib/attribute-transition-manager.js [hunk 3]
```diff
@@ -27,7 +27,7 @@ export default class AttributeTransitionManager {
 
     if (Transform.isSupported(gl)) {
       this.isSupported = true;
-    } else {
+    } else if (gl) {
```
Comment (ibgreen) [inline modules/core/src/lib/attribute-transition-manager.js]: Maybe add a short comment reminding the maintainer that this class must be instantiatable without a gl context?

File: modules/core/src/lib/attribute.js [hunk 4]
```diff
@@ -218,6 +220,9 @@ export default class LayerAttribute extends Attribute {
           this.update({constant: false, value: buffer});
           state.needsRedraw = true;
         }
+      } else {
+        this.update(buffer);
```
Comment (ibgreen) [inline modules/core/src/lib/attribute.js]: Getting a little lost here. What is the type of the buffer in this case? Why are we adding this, typed arrays should be fine on workers? 

Maybe a comment could help?

File: modules/core/src/lib/attribute.js [hunk 5]
```diff
@@ -149,10 +149,12 @@ export default class LayerAttribute extends Attribute {
     if (update) {
       // Custom updater - typically for non-instanced layers
       update.call(context, this, {data, props, numInstances});
-      this.update({
-        value: this.value,
-        constant: this.constant
-      });
+      if (this.gl) {
```
Comment (ibgreen) [inline modules/core/src/lib/attribute.js]: * Would it be more readable if we push the `if (gl)` into the `update` method to reduce code length and logic here?
* Any way to make the responsibility of `update` clearer through renaming (`updateWebGLBuffer` etc)?

File: modules/core/src/lib/deck.js (hunk 1 above)
Comment (ibgreen) [reply modules/core/src/lib/deck.js]: > This is rather unfortunate - when using the browser field in package.json to support node, it does not cover the worker case

Hmm, any way we can handle this inside mjolnir.js? In general, I think we want to strive for all of our libraries to be importable in all targets (browser, node, worker, ...). And the conditional reduction of functionality should ideally be done inside each library, rather than at import.

File: modules/core/src/lib/layer.js [hunk 6]
```diff
@@ -472,16 +472,16 @@ export default class Layer extends Component {
...
-      return this.state.numInstances;
-    }
-
     // Check if app has provided an explicit value
     if (props.numInstances !== undefined) {
       return props.numInstances;
     }
 
+    // First check if the layer has set its own value
+    if (this.state && this.state.numInstances !== undefined) {
```
Comment (ibgreen) [inline modules/core/src/lib/layer.js]: Change is fine but as a side note: `numInstances` semantics feels rather shaky to me. I am not sure I can explain to a user exactly how this field is supposed to work. Might be good to brainstorm and change or at least update docs to cover current usage.

File: modules/layers/src/path-layer/path-layer.js [hunk 7]
```diff
@@ -120,7 +120,7 @@ export default class PathLayer extends Layer {
       if (this.state.model) {
         this.state.model.delete();
       }
-      this.setState({model: this._getModel(gl)});
+      this.setState({model: gl && this._getModel(gl)});
```
Comment (ibgreen) [inline modules/layers/src/path-layer/path-layer.js]: More readable/likely to avoid issues in other places if we push the gl check inside `_getModel`?
 
If `_getModel` is the one sublayers override maybe call it through `getModel` and check there?

File: modules/layers/src/solid-polygon-layer/solid-polygon-layer.js [hunk 8]
```diff
@@ -79,7 +79,13 @@ export default class SolidPolygonLayer extends Layer {
 
     /* eslint-disable max-len */
     attributeManager.add({
-      indices: {size: 1, isIndexed: true, update: this.calculateIndices, noAlloc},
+      indices: {
+        size: 1,
+        isIndexed: true,
+        type: GL.UNSIGNED_INT,
```
Comment (ibgreen) [inline modules/layers/src/solid-polygon-layer/solid-polygon-layer.js]: Why was this needed? Is specifying it here like this compatible with the 16 bit case? Short comment?

File: test/apps/attribute-worker/app.js [hunk 9]
```diff
@@ -0,0 +1,78 @@
+import React, {Component} from 'react';
+import {render} from 'react-dom';
+import DeckGL from 'deck.gl';
+import * as Layers from '@deck.gl/layers';
+import createWorker from 'webworkify-webpack';
+
+import TEST_CASES from './test-cases';
+
+// Set your mapbox token here
+const MAPBOX_TOKEN = process.env.MapboxAccessToken; // eslint-disable-line
```
Comment (ibgreen) [inline test/apps/attribute-worker/app.js]: ? Forgot to remove?

File: test/apps/attribute-worker/app.js [hunk 10]
```diff
@@ -0,0 +1,78 @@
+import React, {Component} from 'react';
```
Comment (ibgreen) [inline test/apps/attribute-worker/app.js]: React seems like unnecessary overhead in a worker, use pure JS version?

File: modules/layers/src/solid-polygon-layer/solid-polygon-layer.js (hunk 8 above)
Comment (Pessimistress) [reply modules/layers/src/solid-polygon-layer/solid-polygon-layer.js]: If `type` is not specified, AttributeManager forces the external buffer to be Float32Array.

File: modules/core/src/lib/layer.js [hunk 11]
```diff
@@ -472,16 +472,16 @@ export default class Layer extends Component {
...
-    if (this.state && this.state.numInstances !== undefined) {
-      return this.state.numInstances;
-    }
-
     // Check if app has provided an explicit value
     if (props.numInstances !== undefined) {
       return props.numInstances;
     }
 
+    // First check if the layer has set its own value
```
Comment (ibgreen) [inline modules/core/src/lib/layer.js]: Change comment: // Second, check if...

File: modules/core/src/lib/attribute.js [hunk 12]
```diff
@@ -206,18 +208,26 @@ export default class LayerAttribute extends Attribute {
...
+      } else if (this.value !== buffer) {
         if (state.auto && buffer.length <= numInstances * this.size) {
           throw new Error('Attribute prop array must match length and size');
         }
-        if (this.value !== buffer) {
+
+        const ArrayType = glArrayFromType(this.type || GL.FLOAT);
+        if (!(buffer instanceof ArrayType)) {
+          log.warn(
+            `Attribute ${this.id} is type ${buffer.constructor.name}, expecting type ${
```
Comment (ibgreen) [inline modules/core/src/lib/attribute.js]: These are pretty extensive error messages, normally we don't do this due to e.g. bundle size concerns from the strings etc/

Maybe it is justified here though?

Also `constructor.name` gets clobbered during minification so this would generated random message in that case, better avoid.

File: modules/core/src/lib/attribute.js [hunk 13]
```diff
@@ -206,18 +208,26 @@ export default class LayerAttribute extends Attribute {
           this.update({constant: false, buffer});
           state.needsRedraw = true;
         }
-      } else {
-        const ArrayType = glArrayFromType(this.type || GL.FLOAT);
-        if (!(buffer instanceof ArrayType)) {
-          throw new Error(`Attribute ${this.id} must be of type ${ArrayType.name}`);
-        }
+      } else if (this.value !== buffer) {
```
Comment (ibgreen) [inline modules/core/src/lib/attribute.js]: Looks like `setExternalBuffer` now takes arrays and not just buffers? ...

Maybe break this part into a separate function and consider naming to help tame complexity/guide maintainers? 

I'm surprised the linter doesn't complain about max-statements, maybe we have some linter issue?

File: modules/core/src/lib/layer-manager.js [hunk 14]
```diff
@@ -88,7 +88,7 @@ export default class LayerManager {
 
       gl,
       // Enabling luma.gl Program caching using private API (_cachePrograms)
-      shaderCache: new ShaderCache({gl, _cachePrograms: true}),
+      shaderCache: gl && new ShaderCache({gl, _cachePrograms: true}),
```
Comment (ibgreen) [inline modules/core/src/lib/layer-manager.js]: If we are going to make certain classes work without WebGL, to make sure we don't regress, we should ideally at least have test cases that instantiate them without WebGL context.

File: modules/layers/src/solid-polygon-layer/polygon-tesselator.js [hunk 15]
```diff
@@ -42,7 +42,7 @@ export default class PolygonTesselator extends Tesselator {
         positions: {size: 3},
         positions64xyLow: {size: 2, fp64Only: true},
         vertexValid: {type: Uint8ClampedArray, size: 1},
-        indices: {type: IndexType, size: 1}
+        indices: {type: Uint32Array, size: 1}
```
Comment (ibgreen) [inline modules/layers/src/solid-polygon-layer/polygon-tesselator.js]: Can't recall, is this now automatically handled on lower level (luma.gl attribute?), or are we dropping support for browsers that don't have the extension?

File: modules/core/src/lib/attribute.js (hunk 12 above)
Comment (Pessimistress) [reply modules/core/src/lib/attribute.js]: Since `buffer` is a typed array constructor.name should not be mangled

File: modules/core/src/lib/attribute.js (hunk 13 above)
Comment (Pessimistress) [reply modules/core/src/lib/attribute.js]: I'm fine with adding a check for typed array here.

This change is to avoid throwing an error if `Uint8Array`/`Float32Array` is provided instead of `Uint8ClampedArray` for colors, or `Uint16Array` is provided instead of `Uint32Array` for indices. The old treatment did not work well as we try to decouple loading from the layer itself.
